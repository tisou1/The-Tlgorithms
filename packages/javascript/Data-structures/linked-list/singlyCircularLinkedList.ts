import { Node } from './singlyLinkedList'
import type { NodeT } from './singlyLinkedList'

class SinglyCircularLinkedList<T = unknown> {
  headNode: NodeT<T> | null 
  length: number
  constructor() {
    this.headNode = null
    this.length = 0
  }

  size = () => this.length

  head = () => this.headNode?.data || null

  isEmpty = () => this.size() === 0

  initiateNodeAndIndex() {
    return {
      currentNode: this.headNode,
      currentIndex: 0
    }
  }

  //获取元素
  getElementAt(index: number) {
    if(this.length !== 0 && index >= 0 && index <= this.length){
      let { currentNode } = this.initiateNodeAndIndex()
      //可以只用index来作为判断依据
      for(let i=0; i< index && currentNode!==null; i++){
        currentNode = currentNode.next
      }
      return currentNode
    }

    return undefined
  }
  //头部插入元素
  addAtFirst(data: T) {
    const node = new Node(data)
    node.next = this.headNode
    this.headNode = node
    this.length++

    return this.length
  }

  //尾部插入
  addAtLast(data: T) {
    //空的,就直接在头部进行插入
    if(!this.headNode) {
      return this.addAtFirst(data)
    }

    const node = new Node(data)

    //获取最后一个元素
    const currentNode = this.getElementAt(this.length - 1)
    currentNode!.next = node
    node.next = this.headNode
    this.length++
    return this.length
  }

  insertAt(index: number, data: T) {
    if(index === 0) {
      return this.addAtFirst(data)
    }
    if(index === this.length){
      return this.addAtLast(data)
    }
    if(index < 0 || index > this.length)
      throw new RangeError('越界了')

    //生成节点
    const node = new Node(data)
    const previousNode = this.getElementAt(index - 1)
    node.next = previousNode!.next
    previousNode!.next = node
    this.length++
    return this.length
  }

  indexOf(data: T) {
    let { currentNode, currentIndex } = this.initiateNodeAndIndex()

    while(currentNode) {
      if(currentNode.data === data) {
        return currentIndex
      }
      currentIndex++

      currentNode = currentNode.next
    }

    return -1
  }

  //尾部移除
  remove () {
    if(this.isEmpty()) return null
    if(this.length === 1) {
      let data = this.headNode?.data
      this.headNode = null
      this.length--
      return data || null
    }
    const secondLastNode = this.getElementAt(this.length - 2)
    const removeedNode = secondLastNode?.next
    secondLastNode!.next = this.headNode
    this.length--
    return removeedNode?.data || null
  }
  //头部移除
  removeFirst() {
    if (this.isEmpty()) return null
    const removedNode = this.headNode
    if (this.length === 1) {
      this.clear()
      return removedNode!.data
    }
    const lastNode = this.getElementAt(this.length - 1)
    this.headNode = this.headNode!.next
    lastNode!.next = this.headNode
    this.length--
    return removedNode!.data || null
  }

  removeAt(index) {
    if(this.isEmpty()) return null
    if(index === 0) return this.removeFirst()
    if(index === this.length) return this.remove()
    if(index < 0 && index > this.length) return null

    const previousNode = this.getElementAt(index - 1)
    const currentNode = previousNode?.next
    previousNode!.next = currentNode!.next
    this.length--
    return currentNode?.data || null
  }

  removeData(data: T) {
    if(this.isEmpty()) return null
    const index = this.indexOf(data)
    return this.removeAt(index)
  }

  printData(output = (value: any) => console.log(value)) {
    let {currentIndex, currentNode } = this.initiateNodeAndIndex()

    while(currentIndex < this.length) {
      output(currentNode?.data)
      //指针后移
      currentNode = currentNode!.next
      currentIndex++
    }
  }

  get() {
    let { currentIndex, currentNode } = this.initiateNodeAndIndex()
    const list: T[] = []
    while(currentIndex < this.length && currentNode) {
      list.push(currentNode!.data)
      currentNode = currentNode!.next
      currentIndex++
    }

    return list
  }


  clear() {
    this.length = 0
    this.headNode = null
  }
}

export { SinglyCircularLinkedList }