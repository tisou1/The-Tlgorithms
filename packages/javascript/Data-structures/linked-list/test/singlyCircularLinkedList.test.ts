import { SinglyCircularLinkedList }  from '../singlyCircularLinkedList'
import { describe, it, expect, beforeEach} from 'vitest'


describe('SinglyCircularLinkedList', () => {
  let list
  //当前上下文中的每个测试运行前被调用
  beforeEach(() => {
    list = new SinglyCircularLinkedList()
  })

  it('Check get', () => {
    expect(list.get()).toEqual([])
    expect(list.addAtFirst(1)).toEqual(1)
    expect(list.addAtFirst(2)).toEqual(2)
    expect(list.get()).toMatchSnapshot()
    expect(list.size()).toBe(2)
    expect(list.get()).toEqual([2,1])
  })

  it('check size', () => {
    expect(list.size()).toEqual(0)
    expect(list.addAtLast(1)).toEqual(1)
    expect(list.addAtLast(2)).toEqual(2)
    expect(list.size()).toBe(2)
  })

  it('check head', () => {
    expect(list.head()).toEqual(null)
    expect(list.addAtLast(1111)).toBe(1)
    expect(list.head()).toEqual(1111)
    list.clear()
    expect(list.head()).toEqual(null)
  })

  it('Check isEmpty', () => {
    expect(list.isEmpty()).toEqual(true)
    expect(list.addAtFirst('123')).toEqual(1)
    expect(list.isEmpty()).toEqual(false)
  })

  it('check getElementAt', () => {
    list.addAtLast(100)
    list.addAtLast(200)
    list.addAtLast(300)
    list.addAtLast(500)
    list.addAtLast(900)

    expect(list.getElementAt(1).data).toEqual(200)
    expect(list.getElementAt(2).data).toEqual(300)
  })

  it('check addAtFitst', () => {
    list.addAtLast(100)
    list.addAtLast(200)
    list.addAtLast(300)

    expect(list.get()).toEqual([100, 200, 300])
    list.addAtFirst(50)
    expect(list.get()).toEqual([50, 100, 200, 300])
  })

  it('check insert', () => {
    expect(list.insertAt(0, 100)).toEqual(1)
    expect(list.insertAt(0, 200)).toEqual(2)
    expect(list.insertAt(2, 300)).toEqual(3)

    expect(list.get()).toEqual([200, 100, 300])
  })

  it('check indexof', () => {
    list.addAtLast(100)
    list.addAtLast(200)
    list.addAtLast(300)

    expect(list.indexOf(200)).toBe(1)
    expect(list.indexOf(300)).toBe(2)
  })

  
})
