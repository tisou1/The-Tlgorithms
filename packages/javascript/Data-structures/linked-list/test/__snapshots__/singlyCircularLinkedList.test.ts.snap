// Vitest Snapshot v1

exports[`SinglyCircularLinkedList > Check get 1`] = `
[
  2,
  1,
]
`;
